source/ada/lsp-ada_handlers.adb

```diff
@@ -142,39 +237,29 @@ package body LSP.Ada_Handlers is
          return Response;
       end if;
 
-      Definition := LSP.Lal_Utils.Resolve_Name (Name_Node);
+      --  Check is we are on some defining name
+      Definition := LSP.Lal_Utils.Get_Name_As_Defining (Name_Node);
 
       if Definition = No_Defining_Name then
-         return Response;
-      end if;
-
-      declare
-
-         use Libadalang.Common;
+         Definition := LSP.Lal_Utils.Resolve_Name (Name_Node);
 
-         Start_Sloc_Range :
-         constant Langkit_Support.Slocs.Source_Location_Range :=
-           Sloc_Range (Data (Definition.Token_Start));
-         End_Sloc_Range   :
-         constant Langkit_Support.Slocs.Source_Location_Range :=
-           Sloc_Range (Data (Definition.Token_End));
-
-         First_Position : constant LSP.Messages.Position :=
-           (Line_Number (Start_Sloc_Range.Start_Line) - 1,
-            UTF_16_Index (Start_Sloc_Range.Start_Column) - 1);
-         Last_Position  : constant LSP.Messages.Position :=
-           (Line_Number (End_Sloc_Range.End_Line) - 1,
-            UTF_16_Index (End_Sloc_Range.End_Column) - 1);
+         if Definition /= No_Defining_Name then
+            Append (Response.result, Definition);
+         end if;
+      else  --  If we are on a defining_name already
+         Other_Part := Find_Next_Part (Definition);
 
-         Location : constant LSP.Messages.Location :=
-           (uri  => Self.Context.File_To_URI (+Definition.Unit.Get_Filename),
-            span => LSP.Messages.Span'(First_Position, Last_Position));
+         if Other_Part = No_Defining_Name then
+            --  No next part is found. Check first defining name
+            Other_Part := Find_First_Part (Definition);
+         end if;
 
-      begin
-         Response.result.Append (Location);
-         return Response;
-      end;
+         if Other_Part /= No_Defining_Name then
+            Append (Response.result, Other_Part);
+         end if;
+      end if;
 
+      return Response;
```

Comment:
It would be great to have tests for the new behaviors

Reply:
done